rust/worker/src/index/hnsw.rs

```diff
@@ -208,18 +208,35 @@ impl Index<HnswIndexConfig> for HnswIndex {
         unsafe { mark_deleted(self.ffi_ptr, id) }
     }
 
-    fn query(&self, vector: &[f32], k: usize) -> (Vec<usize>, Vec<f32>) {
+    fn query(
+        &self,
+        vector: &[f32],
+        k: usize,
+        allowed_ids: &[usize],
```

Comment:
rust slices store the length of the data they refer to, so the index layer should not need to expose this. The FFI call on 225 can do allowed_ids.len()